File: android/src/main/kotlin/uz/tbm/antifraud_flutter/antifraud_flutter/AntifraudFlutterPlugin.kt
package uz.tbm.antifraud_flutter.antifraud_flutter

import android.content.Context
import android.util.Log
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import uz.tbm.antifraudmobilesdk.AntiFraudLibrary

/** AntifraudFlutterPlugin */
class AntifraudFlutterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var context: Context
    private lateinit var channel: MethodChannel
    private var sdk: AntiFraudLibrary? = null // SDK instance

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "antifraud_flutter")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "init" -> {
                val host = call.argument<String>("host") ?: ""
                sdk = AntiFraudLibrary(
                    host = host,
                    context = context,
                )
            }

            "initialize" -> {
                initializeSDK(result)
            }

            "is_initialized" -> {
                val isInitialized = isSDKInitialized()
                result.success(isInitialized);
            }

            "get_client_instance_id" -> {
                getClientInstanceId(result)
            }

            "verify_sms_code" -> {
                val code = call.argument<String>("code") ?: ""
                val phoneNumber = call.argument<String>("phone_number") ?: ""
                verifySMSCode(code, phoneNumber, result)
            }

            "confirm_face" -> {
                val document = call.argument<String>("document") ?: ""
                val birthDate = call.argument<String>("birth_date") ?: ""
                confirmFace(document, birthDate, result)
            }

            "detect_fraud" -> {
                val code = call.argument<String>("code") ?: ""
                detectFraud(code, result)
            }

            "make_operation" -> makeOperation(result)

            "logout" -> logout(result)

            else -> result.notImplemented()
        }
    }

    private fun isSDKInitialized(): Boolean {
        return sdk?.isInitialized() ?: false;
    }

    private fun initializeSDK(
        result: MethodChannel.Result
    ) {
        try {
            sdk?.initialize { r ->
                r.onSuccess {
                    Log.d("AntifraudFlutterPlugin", "SDK initialized successfully")
                    result.success("SDK initialized successfully")
                }
                r.onFailure {
                    Log.e("AntifraudFlutterPlugin", "SDK initialization failed: ${it.message}")
                    result.error("INIT_ERROR", it.message, null)
                }
            }
        } catch (e: Exception) {
            Log.e("AntifraudFlutterPlugin", "Error initializing SDK: ${e.message}")
            result.error("INIT_ERROR", e.message, null)
        }
    }

    private fun verifySMSCode(code: String, phoneNumber: String, result: MethodChannel.Result) {
        if (!isSDKInitialized()) {
            result.error("INIT_ERROR", "SDK is not initialized", null)
            return
        }
        sdk!!.verifySmsCode(code = code, phoneNumber = phoneNumber) { r ->
            r.onSuccess {
                Log.d("AntifraudFlutterPlugin", "SMS code verified successfully")
                result.success("Code verified successfully")
            }
            r.onFailure {
                Log.e("AntifraudFlutterPlugin", "SMS verification failed: ${it.message}")
                result.error("VERIFY_ERROR", it.message, null)
            }
        }
    }


    private fun confirmFace(document: String, birthDate: String, result: MethodChannel.Result) {
        if (!isSDKInitialized()) {
            result.error("INIT_ERROR", "SDK is not initialized", null)
            return
        }
        sdk!!.confirmFace(document = document, birthDate = birthDate) { r ->
            r.onSuccess {
                Log.d("AntifraudFlutterPlugin", "Confirm face successfully")
                result.success("Confirm face successfully")
            }
            r.onFailure {
                Log.e("AntifraudFlutterPlugin", "Confirm face verification failed: ${it.message}")
                result.error("VERIFY_ERROR", it.message, null)
            }
        }
    }

    private fun detectFraud(code: String, result: MethodChannel.Result) {
        if (!isSDKInitialized()) {
            result.error("INIT_ERROR", "SDK is not initialized", null)
            return
        }
        sdk!!.detectFraud(code = code) { r ->
            r.onSuccess {
                Log.d("AntifraudFlutterPlugin", "Detect fraud successfully")
                result.success("Detect fraud successfully")
            }
            r.onFailure {
                Log.e("AntifraudFlutterPlugin", "SMS verification failed: ${it.message}")
                result.error("DETECT_FRAUD_ERROR", it.message, null)
            }
        }
    }

    private fun logout(result: MethodChannel.Result) {
        if (!isSDKInitialized()) {
            result.error("INIT_ERROR", "SDK is not initialized", null)
            return
        }
        sdk!!.logout { r ->
            r.onSuccess {
                Log.d("AntifraudFlutterPlugin", "Logout successfull")
                result.success("Logout successful")
            }
            r.onFailure {
                Log.e("AntifraudFlutterPlugin", "Logout failed: ${it.message}")
                result.error("LOGOUT_ERROR", it.message, null)
            }
        }
    }

    private fun makeOperation(result: MethodChannel.Result) {
        if (!isSDKInitialized()) {
            result.error("INIT_ERROR", "SDK is not initialized", null)
            return
        }
        sdk!!.makeOperation { r ->
            r.onSuccess {
                Log.d("AntifraudFlutterPlugin", "Make Operation successfull")
                result.success("Make Operation successful")
            }
            r.onFailure {
                Log.e("AntifraudFlutterPlugin", "Make Operation failed: ${it.message}")
                result.error("MAKE_OPERATION_ERROR", it.message, null)
            }
        }
    }

    private fun getClientInstanceId(result: MethodChannel.Result): Unit {
        if (!isSDKInitialized()) {
            result.error("INIT_ERROR", "SDK is not initialized", null)
            return
        }
        sdk!!.getClientInstanceId { r ->
            r.onSuccess {
                Log.d("AntifraudFlutterPlugin", "Get Client InstanceId successfull")
                result.success(it)
            }
            r.onFailure {
                Log.e("AntifraudFlutterPlugin", "Make Operation failed: ${it.message}")
                result.error("MAKE_OPERATION_ERROR", it.message, null)
            }
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: ios/Classes/AntifraudFlutterPlugin.swift
import AntiFraud_Mobile_SDK
import Flutter
import UIKit

public class AntifraudFlutterPlugin: NSObject, FlutterPlugin {
    private var library: AntiFraudLibrary?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "antifraud_flutter", binaryMessenger: registrar.messenger())
        let instance = AntifraudFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(
        _ call: FlutterMethodCall, result: @escaping FlutterResult
    ) {
        switch call.method {
        case "init":
            guard let args = call.arguments as? [String: Any],
                  let host = args["host"] as? String
            else {
                result(
                    FlutterError(
                        code: "INVALID_ARGUMENTS",
                        message: "Invalid arguments passed", details: nil))
                return
            }
            library = AntiFraudLibrary(host: host)
            return

        case "initialize":
            initialize(result: result)
            return

        case "is_initialized":
            guard let args = call.arguments as? [String: Any],
                  let host = args["host"] as? String
            else {
                result(
                    FlutterError(
                        code: "INVALID_ARGUMENTS",
                        message: "Invalid arguments passed", details: nil))
                return
            }
            isInitialized(result: result)
            return
        case "get_client_instance_id":
            getClientInstanceId(result: result)
            return
        case "verify_sms_code":
            guard let args = call.arguments as? [String: Any],
                  let phoneNumber = args["phone_number"] as? String,
                  let smsCode = args["code"] as? String
            else {
                result(
                    FlutterError(
                        code: "INVALID_ARGUMENTS",
                        message: "Invalid arguments passed", details: nil))
                return
            }
            verifySmsCode(
                phoneNumber: phoneNumber, smsCode: smsCode, result: result)
            return
        case "detect_fraud":
            guard let args = call.arguments as? [String: Any],
                  let smsCode = args["code"] as? String
            else {
                result(
                    FlutterError(
                        code: "INVALID_ARGUMENTS",
                        message: "Invalid arguments passed", details: nil))
                return
            }
            detectFraud(smsCode: smsCode, result: result)
            return

        case "make_operation":
            makeOperation(result: result)
            return
        case "confirm_face":
            guard let args = call.arguments as? [String: Any],
                  let document = args["document"] as? String,
                  let birthDate = args["birth_date"] as? String
            else {
                result(
                    FlutterError(
                        code: "INVALID_ARGUMENTS",
                        message: "Invalid arguments passed", details: nil))
                return
            }
            confirmFace(birthDate: birthDate, document: document, result: result)
            return
        case "logout":
            logout(result: result)
            return
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initialize( result: @escaping FlutterResult) {
        library?.initialization { response in
            switch response {
            case .success:
                result("SDK Initialized successfully")
            case .failure(let error):
                result(
                    FlutterError(
                        code: "\(error.code)", message: error.message,
                        details: nil))
            @unknown default:
                result(
                    FlutterError(
                        code: "INIT_ERROR", message: "BASE_ERROR", details: nil)
                )
            }
        }
    }

    private func isInitialized(host: String, result: @escaping FlutterResult) {
        let isInitialized = library?.isInitialized()
        result(isInitialized ?? false)
    }

    private func verifySmsCode(
        phoneNumber: String, smsCode: String, result: @escaping FlutterResult
    ) {
        guard let library = library else {
            result(
                FlutterError(
                    code: "INIT_ERROR", message: "SDK is not initialized",
                    details: nil))
            return
        }
        library.verifySmsCode(phoneNumber: phoneNumber, smsCode: smsCode) {
            response in
            switch response {
            case .success:
                result("Verified successfully")
            case .failure(let error):
                result(
                    FlutterError(
                        code: "\(error.code)", message: error.message,
                        details: nil))
            @unknown default:
                result(
                    FlutterError(
                        code: "INIT_ERROR", message: "BASE_ERROR", details: nil)
                )
            }

        }
    }

    private func getClientInstanceId(result: @escaping FlutterResult) {
        guard let library = library else {
            result(
                FlutterError(
                    code: "INIT_ERROR", message: "SDK is not initialized",
                    details: nil))
            return
        }
        do {
            let clientInstanceId = try library.getClientInstanceId()
            result(clientInstanceId)
        } catch {
            result(FlutterError(code: "GET_CLIENT_INSTANCE_ID_ERROR",
                                message: "Failed to get clientInstanceId",
                                details: error.localizedDescription))
        }
    }

    private func confirmFace(
        birthDate: String, document: String, result: @escaping FlutterResult
    ) {
        guard let library = library else {
            result(
                FlutterError(
                    code: "INIT_ERROR", message: "SDK is not initialized",
                    details: nil))
            return
        }
        library.confirmFace(birthDate: birthDate, document: document) {
            response in
            switch response {
            case .success:
                result("Confirmed successfully")
            case .failure(let error):
                result(
                    FlutterError(
                        code: "\(error.code)", message: error.message,
                        details: nil))
            @unknown default:
                result(
                    FlutterError(
                        code: "INIT_ERROR", message: "BASE_ERROR", details: nil)
                )
            }

        }
    }

    private func detectFraud(smsCode: String, result: @escaping FlutterResult) {
        guard let library = library else {
            result(
                FlutterError(
                    code: "INIT_ERROR", message: "SDK is not initialized",
                    details: nil))
            return
        }
        library.detectFraud(smsCode: smsCode) { response in
            switch response {
            case .success:
                result("Transaction verified successfully")
            case .failure(let error):
                result(
                    FlutterError(
                        code: "\(error.code)", message: error.message,
                        details: nil))
            @unknown default:
                result(
                    FlutterError(
                        code: "INIT_ERROR", message: "BASE_ERROR", details: nil)
                )
            }
        }
    }

    private func makeOperation(result: @escaping FlutterResult) {
        guard let library = library else {
            result(
                FlutterError(
                    code: "INIT_ERROR", message: "SDK is not initialized",
                    details: nil))
            return
        }
        library.makeOperation { response in
            switch response {
            case .success:
                result("Session id update successfully")
            case .failure(let error):
                result(
                    FlutterError(
                        code: "\(error.code)", message: error.message,
                        details: nil))
            @unknown default:
                result(
                    FlutterError(
                        code: "INIT_ERROR", message: "BASE_ERROR", details: nil)
                )
            }

        }
    }

    private func logout(result: @escaping FlutterResult) {
        guard let library = library else {
            result(
                FlutterError(
                    code: "INIT_ERROR", message: "SDK is not initialized",
                    details: nil))
            return
        }
        library.logout { response in
            switch response {
            case .success:
                result("Logged out successfully")
            case .failure(let error):
                result(
                    FlutterError(
                        code: "\(error.code)", message: error.message,
                        details: nil))
            @unknown default:
                result(
                    FlutterError(
                        code: "INIT_ERROR", message: "BASE_ERROR", details: nil)
                )
            }
        }
    }
}
